lib/phoenix/transports/long_poller.ex

```diff
@@ -79,75 +72,119 @@ defmodule Phoenix.Transports.LongPoller do
   """
-  def longpoll_pid(conn) do
-    case get_session(conn, session_key(conn)) do
-      nil -> :nopid
-      bin ->
-        pid = :erlang.binary_to_term(bin)
-        if Process.alive?(pid) do
-          {:ok, pid}
-        else
-          {:error, :terminated}
+  def verify_longpoll_topic(%Plug.Conn{params: %{"token" => token, "sig" => sig}} = conn) do
+    case verify(conn, token, sig) do
+      {:ok, priv_topic} ->
+        ref = :erlang.make_ref()
+        :ok = subscribe(conn, priv_topic)
+        :ok = broadcast_from(conn, priv_topic, {:subscribe, ref})
+        receive do
+          {:ok, :subscribe, ^ref} -> {:ok, priv_topic}
+        after
+          pubsub_timeout_ms(conn)  -> {:error, :terminated}
         end
+
+      _ -> :notopic
     end
   end
+  def verify_longpoll_topic(_conn), do: :notopic
 
   @doc """
   Ack's a list of `%Phoenix.Socket.Messages{}`'s back to the `Phoenix.LongPoller.Server`
 
   To be called after buffered messages have been relayed to client
   """
-  def ack(server_pid, messages) do
-    :ok = GenServer.call(server_pid, {:ack, messages})
-  end
-
-  defp session_key(conn), do: "#{router_module(conn)}_longpoll_pid"
-
-  @doc """
-  Sets the active listener process. Called by polling `conn`
-  """
-  def set_active_listener(server_pid, listener_pid) do
-    GenServer.call(server_pid, {:set_active_listener, listener_pid})
+  def ack(conn, priv_topic, messages) do
+    ref = :erlang.make_ref()
+    :ok = broadcast_from(conn, priv_topic, {:ack, messages, ref})
+    receive do
+      {:ok, :ack, ^ref} -> :ok
+    after
+      pubsub_timeout_ms(conn) -> :error
+    end
   end
 
   @doc """
   Dispatches deserialized `%Phoenix.Socket.Message{}` from client to
   `Phoenix.LongPoller.Server`
   """
-  def dispatch(server_pid, message) do
-    GenServer.call(server_pid, {:dispatch, message})
+  def dispatch(conn, priv_topic, msg) do
+    ref = :erlang.make_ref()
+    :ok = broadcast_from(conn, priv_topic, {:dispatch, msg, ref})
+    receive do
+      {:ok, :dispatch, ^ref}            -> :ok
+      {:error, :dispatch, reason, ^ref} -> {:error, reason}
+    after
+      pubsub_timeout_ms(conn) -> {:error, :timeout}
+    end
   end
 
   defp timeout_window_ms(conn) do
     get_in endpoint_module(conn).config(:transports), [:longpoller_window_ms]
   end
+
+  defp pubsub_timeout_ms(conn) do
+    get_in endpoint_module(conn).config(:transports), [:longpoller_pubsub_timeout_ms]
+  end
+
+  defp pubsub_server(conn), do: endpoint_module(conn).__pubsub_server__()
+
+  defp subscribe(conn, priv_topic) do
+    Phoenix.PubSub.subscribe(pubsub_server(conn), self, priv_topic, link: true)
+  end
+
+  defp broadcast_from(conn, priv_topic, msg) do
+    Phoenix.PubSub.broadcast_from(pubsub_server(conn), self, priv_topic, msg)
+  end
+
+  defp sign(conn, priv_topic) do
+    salt = derive_salt(conn, to_string(pubsub_server(conn)), [])
+    Plug.Crypto.MessageVerifier.sign(priv_topic, conn.secret_key_base <> salt)
```

Comment:
The secret should be only the salt, including the `conn.secret_key_base` may even be unsafe.